### ISSUE TITLE
Multiple video.js recorder on a single page (iOS)

### ISSUE THREAD

**Issue by tstrohmeier** [collab-project/videojs-record#564]
## Description
In iOS only one recorder at a time can be used.

## Steps to reproduce
Create 2 video.js recorder on a single page. Start the recording in one recorder then in the other recorder. The first recorder is going to continue recording but is black and has no video stream anymore.

### Expected
Use a shared stream in both recorders. "Global Stream" as mentioned here: https://webrtchacks.com/guide-to-safari-webrtc/

### Actual
Only one recorder at a time is usable.

## Additional Information
Please include any additional information necessary here. Including the following:
#### videojs record
4.3.0
#### webm-wasm
0.4.1
#### browsers
iOS Safari v14
#### OSes
iOS

**Comment by thijstriemstra** [comment]
@tstrohmeier what version of iOS?

>  The first recorder is going to continue recording but is black and has no video stream anymore.

This actually sounds like this issue: #552

**Comment by tstrohmeier** [comment]
@thijstriemstra Yes webm-wasm actually works on iOS.  
#552 is the reason why I use webm-wasm. The RecordRTC recorder seems to have an issue on iOS.

But this thicket is about an other problem/topic. iOS only allows one user media stream at a time. It would be good if you can use one stream in multiple  recorder.

**Comment by thijstriemstra** [comment]
> The RecordRTC recorder seems to have an issue on iOS.

No, as you can see in #552 it's Safari that has an issue with MediaRecorder IMO.

> But this thicket is about an other problem/topic. iOS only allows one user media stream at a time.

No I think it's the same issue. One user media stream in iOS is causing behavior your see in #552

> It would be good if you can use one stream in multiple recorder.

It would be good but I don't see how to fix this, without having to make all kinds of workarounds just for Safari. I don't get paid for this work and spending all my time on Safari is just plain annoying, Apple isn't paying me or something. If it works correctly in all browsers, I would expect it to work in Safari as well, and if it doesn't it's a bug in Safari. As you can see in #552, the latest Safari preview release has fixed the issue and I am certain it will be fixed in iOS as well. Until then I welcome any pull requests but you might as well wait until Safari releases a new version and fix for this issue.

**Comment by tstrohmeier** [comment]
> No, as you can see in #552 it's Safari that has an issue with MediaRecorder IMO. 

I did not debug on iOS yet but on MacOS Safari v14 I can confirm that MediaRecorder does not exists:
`ReferenceError: Can't find variable: MediaRecorder`

> webm-wasm plugin also uses RecordRTC.

I don't know. I just use the library. The benefit for me is that I can also control bit rate etc.


> No I think it's the same issue. One user media stream in iOS is causing behavior your see in #552

Not really. As I wrote the webm-wasm plugin **works** for me in all browsers (Chrome, Firefox, Safari, Safari iOS). 

The only issue I have is that iOS only allows one stream at the time and so I can only use one recorder at the time on iOS/iPadOS. This is expected behavior on iOS. 
Here is a good description on it: https://webrtchacks.com/guide-to-safari-webrtc/  Headline: "_New getUserMedia() request kills existing stream track_" I also don't think that Apple will change this. 

> It would be good but I don't see how to fix this, without having to make all kinds of workarounds just for Safari. I don't get paid for this work and spending all my time on Safari is just plain annoying, Apple isn't paying me or something. If it works correctly in all browsers, I would expect it to work in Safari as well, and if it doesn't it's a bug in Safari. As you can see in #552, the latest Safari preview release has fixed the issue and I am certain it will be fixed in iOS as well. Until then I welcome any pull requests but you might as well wait until Safari releases a new version and fix for this issue.

I can understand you! I do not have many experience with WebRTC so I can not really help you. 
Seems like Apple is doing it's own thing again. I just notice from reading a bit that iOS will never change this behavior and this is why the possibility of getter/setter for the underlying media stream would be a good thing for the future. This would allow to share it between different  recorder instances.

**Comment by thijstriemstra** [comment]
> I did not debug on iOS yet but on MacOS Safari v14 I can confirm that MediaRecorder does not exists:
ReferenceError: Can't find variable: MediaRecorder

It should be enabled by default. If not, use these instructions to enable it: https://www.educative.io/edpresso/how-to-enable-the-mediarecorder-api-for-safari But it's buggy for now: #552

**Comment by thijstriemstra** [comment]
I tested this with the [updated](https://github.com/collab-project/videojs-record/commit/9517311688247533b1051554ff3b5946953e2e58) multi example on iOS 14.5 on Mobile Safari 14.1 and when I start the video preview in the 2nd recorder it goes black like you described. It seems to work fine on desktop Safari 14.1.

The related webkit ticket is https://bugs.webkit.org/show_bug.cgi?id=179363 so I'm marking this as a known browser issue.
